feat: make gRPC timeout configurations user-configurable

CuteChuanChuan opened the pull request:
# Which issue does this PR close?
Closes #117

# Rationale for this change
PR #115 introduced gRPC timeout configurations, but these values were hard-coded. This PR extends #115 by making all gRPC timeout values user-configurable through DataFusion's extensible configuration mechanism.

# What changes are included in this PR?
- Added 9 new configuration options to `BallistaConfig` (5 client-side, 4 server-side)
- Modified `create_grpc_client_connection()` and `create_grpc_server()` to accept config parameter
- Updated all call sites across executor and scheduler

# Are there any user-facing changes?
**Yes.** Users can now configure gRPC timeouts.

All existing tests pass.

milenkovicm commented on ballista/core/src/execution_plans/distributed_query.rs:
```diff
@@ -238,6 +238,7 @@ impl<T: 'static + AsLogicalPlan> ExecutionPlan for DistributedQueryExec<T> {
                 self.session_id.clone(),
                 query,
                 self.config.default_grpc_client_max_message_size(),
+                self.config.clone(),
```
we already have `gpc_max_message_size` as a parameter can we pass other needed values as parameters rather than passing config

milenkovicm commented on ballista/core/src/execution_plans/distributed_query.rs:
```diff
@@ -274,10 +275,11 @@ async fn execute_query(
     session_id: String,
     query: ExecuteQueryParams,
     max_message_size: usize,
+    config: BallistaConfig,
```
we already have gpc_max_message_size as a parameter can we pass other needed values as parameters rather than passing config

milenkovicm commented on ballista/executor/src/executor_server.rs:
```diff
@@ -114,7 +117,7 @@ pub async fn startup<T: 'static + AsLogicalPlan, U: 'static + AsExecutionPlan>(
         tokio::spawn(async move {
             let shutdown_signal = grpc_shutdown.recv();
-            let grpc_server_future = create_grpc_server()
+            let grpc_server_future = create_grpc_server(&ballista_config)
```
this configuration should be provided as part of executor binary configuration, not ballista condif

milenkovicm commented on ballista/core/src/utils.rs:
```diff
@@ -106,31 +107,50 @@ pub async fn collect_stream(
 }
 
-pub fn create_grpc_server() -> Server {
+pub fn create_grpc_server(config: &BallistaConfig) -> Server {
```
as i mentioned before, IMHO it makes more sense to pass individual parameters, or we can derive some kind of GprcConfig if there are too many of them

CuteChuanChuan replied:
I derive `GrpcServerConfig` for this.

CuteChuanChuan commented:
Hi @milenkovicm ,
Thanks for your comments and pointing out better approach. PTAL when you have a chance.
Thanks.

milenkovicm commented on ballista/client/tests/context_checks.rs:
```diff
@@ -240,12 +240,12 @@ mod supported {
-            "+---------------------------------------+----------+",
+            "+-------------------------------------------------------+-------+",
+            "| name                                                  | value |",
+            "+-------------------------------------------------------+-------+",
```
can we please update query to return ` ballista.job.name` only. so we don't change the test every time we add new option.

milenkovicm commented on ballista/core/src/utils.rs:
```diff
@@ -36,6 +37,64 @@ use std::{fs::File, pin::Pin};
+}
+
+impl GrpcClientConfig {
+    pub fn from_ballista_config(config: &BallistaConfig) -> Self {
```
would it make sense to implement this as `From` or `Into` not sure which one would be better

milenkovicm commented on ballista/scheduler/src/standalone.rs:
```diff
@@ -92,8 +93,9 @@ pub async fn new_standalone_scheduler_with_builder(
+    let grpc_server_config = GrpcServerConfig::default();
     tokio::spawn(
-        create_grpc_server()
+        create_grpc_server(&grpc_server_config)
```
I would suggest

```rust
create_grpc_server(&GrpcServerConfig::default();)
```

if possible

milenkovicm reviewed (commented):
Firs of all sorry for late review, and thank you for your contribution ! 
I think changes make sense, I have just few minor comments, please have a look when you get chance

CuteChuanChuan commented on ballista/core/src/utils.rs:
```diff
@@ -36,6 +37,64 @@ use std::{fs::File, pin::Pin};
 use tonic::transport::{Channel, Error, Server};
 
+#[derive(Debug, Clone)]
+pub struct GrpcClientConfig {
```
Absolutely!

milenkovicm reviewed (approved):
i have just two small coments, if you agree

CuteChuanChuan commented:
Hi @milenkovicm ,

I've addressed the feedback about simplifying the config usage and error found in Clippy. However, I noticed the CI is failing with rustdoc errors in `display.rs` and `execution_graph.rs`, which are files I haven't modified in this PR.

Would you like me to:
  1. Fix them in this PR as well, or
  2. Create a separate issue/PR for these rustdoc fixes?

milenkovicm commented:
thanks on your contribution @CuteChuanChuan
